### PR TITLE
implement Drop Constraint in dedicated request and action

### PR DESCRIPTION
**Pull request by BaurzhanSakhariev**
migrating DROP CONSTRAINT to dedicated delta-only request/action instead of overwriting existing mapping.

### CODE REVIEW

File: server/src/test/java/io/crate/integrationtests/DDLIntegrationTest.java [hunk 1]
```diff
@@ -438,6 +438,15 @@ public void test_alter_table_drop_constraint_removes_the_constraint_and_leaves_o
                          INTERNAL_ERROR,
                          BAD_REQUEST,
                          4000));
+
+        // Remove last constraint
+        execute("alter table t drop constraint check_qty_gt_zero");
+
+        execute(selectCheckConstraintsStmt);
```
Comment (BaurzhanSakhariev) [inline server/src/test/java/io/crate/integrationtests/DDLIntegrationTest.java:445]: Added in order not to go all the way with creating table for dropping one constraint - I think this is just complementary stuff maybe just need to update method name to smth like `can_drop_all_constraints`

File: server/src/test/java/io/crate/integrationtests/PartitionedTableIntegrationTest.java [hunk 2]
```diff
@@ -2364,4 +2364,38 @@ public void test_select_partitioned_by_column_with_query_then_fetch_plan() throw
             "1| Trillian\n"
         ));
     }
+
+    @Test
+    @UseRandomizedSchema(random = false)
+    public void test_alter_table_drop_constraint_on_partitioned_table() throws Exception {
```
Comment (BaurzhanSakhariev) [inline server/src/test/java/io/crate/integrationtests/PartitionedTableIntegrationTest.java:2370]: this is failing on master, so I think makes sense to have dedicated test.

File: server/src/main/java/io/crate/execution/ddl/tables/TransportDropConstraintAction.java [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+                                  DropConstraintRequest request) throws IOException {
+        Index[] concreteIndices = resolveIndices(currentState, request.relationName().indexNameOrAlias());
+        boolean updated = false;
+
+        for (Index index : concreteIndices) {
+            final IndexMetadata indexMetadata = currentState.metadata().getIndexSafe(index);
+
+            Map<String, Object> currentSource = indexMetadata.mapping().sourceAsMap();
+
+            updated |= constraintRemoved(currentSource, request.constraintName());
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/execution/ddl/tables/TransportDropConstraintAction.java:130]: was not sure whether updated = true, when ALL or at least one. For non-partitioned tables it's always one index(?) so it doesn't matter here. 

In general, I added this updated flag to avoid unnecessary 
`ClusterState.builder(currentState).metadata(metadataBuilder).build()` calls but I think we already cut non-existing check scenario in analyzer, so can just remove this flag thingy both from here and above. wdyt?

File: server/src/main/java/io/crate/execution/ddl/tables/TransportDropConstraintAction.java [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+import org.elasticsearch.transport.TransportService;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Map;
+
+import static io.crate.metadata.cluster.AlterTableClusterStateExecutor.resolveIndices;
+
+@Singleton
+public class TransportDropConstraintAction extends AbstractDDLTransportAction<DropConstraintRequest, AcknowledgedResponse> {
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/execution/ddl/tables/TransportDropConstraintAction.java:58]: This action's executor very similar to the executor of the new ADD COLUMN, I can actually create one common executor accepting a mapping mutator function (maybe partition mapping mutator as well) - but I will do it in the PR which gets merged latest as a follow up

File: server/src/main/java/io/crate/planner/node/ddl/AlterTableDropCheckConstraintPlan.java [hunk 5]
```diff
@@ -55,31 +52,18 @@ public void executeOrFail(DependencyCarrier dependencies,
                               RowConsumer consumer,
                               Row params,
                               SubQueryResults subQueryResults) {
-        dependencies.alterTableOperation()
-            .executeAlterTableAddColumn(bind(dropCheckConstraint))
-            .whenComplete(new OneRowActionListener<>(consumer, rCount -> new Row1(rCount == null ? -1 : rCount)));
-    }
+        if (plannerContext.clusterState().getNodes().getMinNodeVersion().before(Version.V_5_1_0)) {
```
Comment (seut) [inline server/src/main/java/io/crate/planner/node/ddl/AlterTableDropCheckConstraintPlan.java:55]: I think we should prevent this in a generic way for all non-DML operations. Maybe hooked up while checking supported table operations https://github.com/crate/crate/blob/master/server/src/main/java/io/crate/metadata/table/Operation.java#L119. (would require passing in the clusterState). As it's not completely straight-forward and may requires discussion, I'd suggest to do this dedicated (follow up). So either we should add an TODO comment here or remove this check.

File: server/src/test/java/io/crate/integrationtests/DDLIntegrationTest.java (hunk 1 above)
Comment (seut) [reply server/src/test/java/io/crate/integrationtests/DDLIntegrationTest.java:445]: Can we move this into a dedicated test case? like `test_can_drop_single_constraint` with a comment explaining that this resulting in an issue before..

File: server/src/test/java/io/crate/integrationtests/PartitionedTableIntegrationTest.java [hunk 6]
```diff
@@ -2364,4 +2364,38 @@ public void test_select_partitioned_by_column_with_query_then_fetch_plan() throw
...
+                "from information_schema.table_constraints " +
+                "where table_name='t'" +
+                "order by constraint_name";
+        execute(selectCheckConstraintsStmt);
+        assertThat(printedTable(response.rows()), is(
+            "doc| t| CHECK| check_id_gt_zero\n" +
+                "doc| t| CHECK| check_x_gt_zero\n" +
+                "doc| t| CHECK| doc_t_id_not_null\n" +
+                "doc| t| PRIMARY KEY| t_pk\n"
+        ));
```
Comment (seut) [inline server/src/test/java/io/crate/integrationtests/PartitionedTableIntegrationTest.java:2389]: I think this is not related to test a successful drop constraint.

File: server/src/test/java/io/crate/integrationtests/PartitionedTableIntegrationTest.java [hunk 7]
```diff
@@ -2364,4 +2364,38 @@ public void test_select_partitioned_by_column_with_query_then_fetch_plan() throw
...
+
+        execute("alter table t drop constraint check_x_gt_zero");
+        refresh();
+
+        execute(selectCheckConstraintsStmt);
```
Comment (seut) [inline server/src/test/java/io/crate/integrationtests/PartitionedTableIntegrationTest.java:2394]: could use a `WHERE constraint_name = 'check_x_gt_zero` filter to be more concrete.

File: server/src/test/java/io/crate/integrationtests/PartitionedTableIntegrationTest.java (hunk 2 above)
Comment (seut) [reply server/src/test/java/io/crate/integrationtests/PartitionedTableIntegrationTest.java:2370]: Can we add a comment that this validates a previous bug?

File: server/src/main/java/io/crate/execution/ddl/tables/TransportDropConstraintAction.java (hunk 3 above)
Comment (seut) [reply server/src/main/java/io/crate/execution/ddl/tables/TransportDropConstraintAction.java:130]: > was not sure whether updated = true, when ALL or at least one. For non-partitioned tables it's always one index(?) so it doesn't matter here.

If at least one changes the state, updated must be true (like your code does). Or do I misunderstand your comment?

Due to concurrent calls, the analyzer state could be different to the actual cluster state. So this code could be called even that the constraint is already remove. So avoiding unnecessary state changed is good to do.

**Review by seut — COMMENTED** [review]
Looks good, added some suggestions.

File: server/src/main/java/io/crate/execution/ddl/tables/TransportDropConstraintAction.java (hunk 3 above)
Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/execution/ddl/tables/TransportDropConstraintAction.java:130]: > at least one

yep, was thinking the same. the only reason about thinking about ALL was that having several indices and not finding constraint in one of them it's not an expected, isn't it? maybe need at least `WARN` about it - but not sure this case is even possible.

File: server/src/test/java/io/crate/integrationtests/DDLIntegrationTest.java (hunk 1 above)
Comment (BaurzhanSakhariev) [reply server/src/test/java/io/crate/integrationtests/DDLIntegrationTest.java:455]: moved in  https://github.com/crate/crate/pull/12996/commits/b2c382f8388700ee12a4389a5acaf41a4a08c00e

File: server/src/main/java/io/crate/planner/node/ddl/AlterTableDropCheckConstraintPlan.java (hunk 5 above)
Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/planner/node/ddl/AlterTableDropCheckConstraintPlan.java]: removed the check - let's do it dedicated

**Comment by BaurzhanSakhariev** [comment]
Hi @seut, thanks for the feedback, pushed fixups.

File: server/src/main/java/io/crate/execution/ddl/tables/DropConstraintRequest.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+
+import javax.annotation.Nonnull;
+import java.io.IOException;
+
+public class DropConstraintRequest extends AcknowledgedRequest<DropConstraintRequest> {
+
+    private final RelationName relationName;
+    private final boolean isPartitioned;
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/execution/ddl/tables/DropConstraintRequest.java:35]: Is it necessary to provide this information? I think it would be possible to derive this based on the RelationName and clusterState

Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/execution/ddl/tables/DropConstraintRequest.java:35]: I tried to remove the flag and compute it as
```
Index[] concreteIndices = resolveIndices(currentState, request.relationName().indexNameOrAlias());
boolean isPartitioned = false;
for (Index index: concreteIndices) {
    if (IndexParts.isPartitioned(index.getName())) {
        isPartitioned = true;
        break;
    }
}
```

but test started to fail as `concreteIndices` was empty - I guess it might work only on non-empty tables but otherwise we need this flag?

Comment (mfussenegger) [reply server/src/main/java/io/crate/execution/ddl/tables/DropConstraintRequest.java:35]: You could check if there's a template in the cluster state that matches the relationName.

The reason I'm asking is also because of https://github.com/crate/crate/issues/11939

Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/execution/ddl/tables/DropConstraintRequest.java]: done in https://github.com/crate/crate/pull/12996/commits/11e662e481a48dca69d5810262083f7637f2b1e2, thanks

File: server/src/main/java/io/crate/planner/node/ddl/AlterTableDropCheckConstraintPlan.java (hunk 5 above)
Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/planner/node/ddl/AlterTableDropCheckConstraintPlan.java:61]: as discussed, made it BWC with TODO
https://github.com/crate/crate/pull/12996/commits/b6ad33d61b95a68a1480c2f19160a1a980bd1d11

File: server/src/test/java/io/crate/integrationtests/PartitionedTableIntegrationTest.java [hunk 9]
```diff
@@ -2364,4 +2366,28 @@ public void test_select_partitioned_by_column_with_query_then_fetch_plan() throw
...
+        refresh();
+
+        String selectCheckConstraintStmt =
+            "select table_schema, table_name, constraint_type, constraint_name " +
+                "from information_schema.table_constraints " +
+                "where table_name='t' and constraint_name = 'check_x_gt_zero' " +
+                "order by constraint_name";
+
+        execute("alter table t drop constraint check_x_gt_zero");
+        refresh();
```
Comment (mfussenegger) [inline server/src/test/java/io/crate/integrationtests/PartitionedTableIntegrationTest.java:2388]:
```suggestion
        String selectCheckConstraintStmt =
            "select table_schema, table_name, constraint_type, constraint_name " +
                "from information_schema.table_constraints " +
                "where table_name='t' and constraint_name = 'check_x_gt_zero' " +
                "order by constraint_name";

        execute("alter table t drop constraint check_x_gt_zero");
```

`refresh` is refreshing the segments/readers, that shouldn't be necessary to observe schema changes.

**Review by mfussenegger — APPROVED** [review]
